datafusion/functions/src/datetime/date_part.rs

```diff
@@ -148,7 +148,9 @@ impl ScalarUDFImpl for DatePartFunc {
     }
 
     fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
-        internal_err!("return_type_from_exprs shoud be called instead")
+        // return type could be Float32 or Int32 depending on the input arguments
+        // return_type_from_exprs should be called instead
+        Ok(DataType::Float64)
```

Comment:
I would appreciate exception more than successfully retrieving a false value.
the false value may easily cause some hard to track down problem, while exception is so much easier to find

Reply:
I have reverted this change now.

Reply:
However, this is a breaking change and is currently blocking Comet from upgrading its DataFusion version. We do not have access to the logical `Expr` to call the newer `return_type_from_exprs` because we are only working with DataFusion's physical plans. I think I would prefer the non-breaking behavior of maintaining the functionality from DataFusion 43 and marking this as deprecated until we can find a better solution.

Maybe @alamb has thoughts on how we can best handle this to avoid causing pain for downstream projects?

Reply:
> this is a breaking change and is currently blocking Comet from upgrading its DataFusion version.

i agree this is breaking and sorry to hear it affects Comet
i wish i knew how to make it non-breaking!

it seems this all boils down to the problem that expressions don't know their types -- https://github.com/apache/datafusion/issues/12604